### PR TITLE
Enhancements for Dockerfile

### PR DESCRIPTION
**Pull request by Phaow**
## Enhancements for Dockerfile
This is an enhancement for dockerfile. 

- Make the comments consist with the steps
- Cache dependencies and code folders( we add new test cases mostly changed is the `test` folder), the other layers will be reused and speed up build and push images when we have the cache configuration

**Test Record(Using a job trigger the test)**
```console
apiVersion: batch/v1
kind: Job
metadata:
  name: kcp-period-test
spec:
  template:
    metadata:
      labels:
        app: kcp-tests
    spec:
      volumes:
        - name: config-data
          configMap:
            name: kcp-tests-config
      containers:
        - name: smoke-tests
          image: 'quay.io/kcp-tests/kcp-tests'
          env:
          - name: KUBECONFIG
            value: "/tmp/kcp-kubeconfig"
          imagePullPolicy: Always
          resources: {}
          command: ["/bin/sh", "-c"]
          args: [ "cp /mnt/config/kcp-kubeconfig /tmp/ && source /mnt/config/set-offline-token && bash /mnt/config/login_with_offline_token.sh && kcp-tests run smoke --include-success || true" ]
          volumeMounts:
            - name: config-data
              mountPath: /mnt/config
          securityContext:
            capabilities:
              drop:
                - ALL
            allowPrivilegeEscalation: false
      restartPolicy: Never
      terminationGracePeriodSeconds: 30
      securityContext:
        runAsNonRoot: true
        seccompProfile:
          type: RuntimeDefault
  backoffLimit: 1

Test result
Failing tests:
[area/apiexports] Author:pewang-Critical-[Smoke] Verify APIBinding working with personal workspace [Suite:kcp/smoke/parallel/minimal]
error: 1 fail, 5 pass, 2 skip (1m39s)
```
**FYI**
The 1 failed is caused by PR https://github.com/kcp-dev/kcp-tests/pull/19. I'll fix it later.

### CODE REVIEW

**Comment by Phaow** [comment]
@xingxingxia @kasturinarra @wangke19 @duanwei33 @zxiao-redhat Could you please help to have a review when you get a chance? Thanks!

**Comment by xingxingxia** [comment]
Great! LGTM
/assign @kasturinarra

**Comment by zxiao-redhat** [comment]
lgtm

File: Dockerfile [hunk 1]
```diff
@@ -16,20 +16,32 @@
...
+COPY bindata.mk bindata.mk
+COPY pkg/ pkg/
+COPY cmd/ cmd/
+COPY hack/ hack/
+COPY test/ test/
+
+# Build the kcp-tests binary
 RUN make build
 
 FROM quay.io/centos/centos:stream8
```
Comment (kasturinarra) [inline Dockerfile:39]: @Phaow should we use more of a RHEL base Image instead of centos ?

Comment (Phaow) [reply Dockerfile:39]: Hi @kasturinarra , we are in an public repo, from my understanding we'd better use a opensource image. RHEL seems for commercial usage. I didn't check more about RHEL docker image, do you have the name+tag?

Comment (kasturinarra) [reply Dockerfile:39]: @Phaow how about this ? https://hub.docker.com/r/redhat/ubi8/tags

Comment (Phaow) [reply Dockerfile:39]: Good suggestion. I think we could consider it. I'll do some test and update later maybe next week plan. Seems we could simply migrate centos os base image to ubi8, tbh, I'll also build the image and test a round.

**Pull request merged**